doc/parameter_server.md

```diff
@@ -0,0 +1,229 @@
+
+Other parameters could be initialized before training.
+
+
+
+## Distribution Strategy
+
+We apply different strategies to different scenarios.
+
+
+| scenario| hardware | update stragety | communication  stragety |
+| :----  |:----  |:---|:---|
+| Recommendation  | CPU|  async SGD|  Parameter server |
+| CV | GPU | BSP SGD| AllReduce|
+| NLP | GPU | BSP SGD| AllReduce |
+
+
+BSP is short for Bulk synchronous parallel.
+
+Following is the assumption under recommendation scenario:
+
+- support asynchronous SGD
+- support elastic scheduling and failover of pserver/worker
+
+Please refer to [Elastic Scheduling](#Elastic Scheduling) part for more details.
+
+
+**Note**
+
+AllReduce failover and pserver failover will be discussed on other design docs.
+
+## Master
+
+- Master is responsible for creating/deleting/scheduling pserver and worker
+- Master defines a parameter sharding strategy, and generate unique key for each parameter(or sharded parameter)
+- Master defines a hash strategy from parameter key to pserver id 
+
+Workflow:
+
+
+1. Master creates several pservers and workers according to user's configuration
+2. Master generates parameter sharding strategy based on model definition provided by user
+3. Master initializes parameter, and send to pservers
+4. Master triggers pservers and workers and start training
+5. Master monitors cluster, and increase or decrease workerers according to priority
+
+
+## Pserver
+
+- Pserver provides kv store service, and workers could push/pull <key, value> to the pserver
+- Pserver provides optimize service, apply gradients to parameters and update back to kv store
+
+
+Workflow:
+
+1. Pserver waits for workers pushing <key, gradient> to pserver
+2. Pserver gets the key and queries kv store to get corresponding parameter
+3. Pserver calls optimizer, applys <key, gradient> to <key, parameter>
+4. Pserver updates <key, parameter> back to kv store
+5. Pserver saves model checkpoint to disk periodially
+
+
+**Note**
+
+We could implement the kv store by ourselves, or we could use some already solution, such as redis.
+
+## Worker
+
+- Worker defines a forward/backward computation proccess
+- Worker defines a dataloader module
+
+Workflow:
+
+1. Before forward layer starting, worker sends key to pserver, and pull <key, parameter> back from pserver
+2. Worker starts forward layer computation
+3. After backward layer finishing, worker generates <key, gradient>
+4. Worker pushes <key, gradient> to pserver
+
+
+There is also another kind of worker who does an evalution job. It define its own evalution computation process and dataloader of validation data. It pulls <key, parameter> from pserver, but never push data to pserver.
```

Comment:
Currently, ElasticDL's worker treats evaluation tasks and training tasks equally. They just request tasks and process tasks. If we add a single validation node, the evaluation speed may suffer.

Reply:
Done